README.md
```
╔═╗╔═╗╔╦╗╔═╗╦ ║╔═╗╔╦╗╔═╗╦╔═║ ╦  ╦  ╔═╗╦ ╦╔═╗╔╦╗╦╔╗║
╚═╗╠═╣║║║║╣ ╠═╣╠═╣ ║║╠═╣╠╩╗║ ║  ║  ║╣ ║║║╠═╣ ║ ║║║║
╚═╝║ ║╩ ╩╚═╝║ ╩║ ║═╩╝║ ║╩ ╩╚═╝  ╚═╝╚═╝╚╩╝╩ ╩ ╩ ╩║╚╝
```

# Samehadaku Lewatin

[![NPM](https://nodei.co/npm/samehadaku-lewatin.png?compact=true)](https://nodei.co/npm/samehadaku-lewatin/)

![GitHub package.json version](https://img.shields.io/github/package-json/v/anasrar/samehadaku-lewatin.svg) ![GitHub](https://img.shields.io/github/license/anasrar/samehadaku-lewatin.svg)
![GitHub last commit](https://img.shields.io/github/last-commit/anasrar/Samehadaku-Lewatin.svg)

🏃‍ Just Bypassing URL Shorter Using Node.js From Samehadaku.tv, Work on Linux, Windows, and Termux

# Install With NPM

```
$ npm install samehadaku-lewatin -g
```

# Use With NPX

```
$ npx samehadaku-lewatin <args>
```

# Install With Git

Clone This Repository

```
$ git clone https://github.com/anasrar/Samehadaku-Lewatin.git samehadaku
```

Change Directory

```
$ cd samehadaku
```

## Install Dependencies

```
$ npm install
```

## Link Into Binary

```
$ npm link
```

# How To Use

Get List Anime From Homepage
```
$ samehadaku list
```

![DEMO](DEMO.gif)

___

Get List Anime From Specific Page
```
$ samehadaku page 3
```

Get List Anime From URL Page
```
$ samehadaku from "https://www.samehadaku.tv/2019/04/fairy-tail-episode-304-subtitle-indonesia.html"
```

Save All Link Quality From Specific Page Use Flag ```-s``` or ```--save```
```
$ samehadaku list -s
```
```
$ samehadaku page 2 --save
```
```
$ samehadaku from "https://www.samehadaku.tv/2019/04/fairy-tail-episode-304-subtitle-indonesia.html" --save
```

Link Save In JSON Format On Some .txt File

## Downloader

### Google Drive

```
samehadaku gdrivedl <url>
```
example

```
samehadaku gdrive "https://drive.google.com/open?id=xxx"
or
samehadaku gdrive "https://drive.google.com/file/d/xxx/view"
```

Thanks To [ariakm25](https://github.com/ariakm25) For Make Great Tools

https://github.com/ariakm25/GDriveDL

### Zippyshare

```
samehadaku zippydl <url>
```
example

```
samehadaku zippydl "https://xxx.zippyshare.com/v/xxx/file.html"
```

https://github.com/anasrar/Zippy-DL

# Changelog
[HERE](CHANGELOG.md)

# License

This project is licensed under the MIT License - see the [LICENSE](LICENSE) file for details

CHANGELOG.md
# Changelog
## 3.0.1 - 05/09/2019
### Added
- #4 New Safelink URL [ ahexa.com ]

## 3.0.0 - 26/08/2019
### Update
- Use Async For Better Performance
- Search Command Removed
### Added
- #4 New Safelink URL [ anjay.info ]


## 2.4.3 - 08/08/2019
### Update
- Search Command Deprecated
- Fix From Command

## 2.4.2 - 28/07/2019
### Update
- Fix Tags Post Filter
### Added
- New Safelink URL [ eue.siherp.com ]

## 2.4.1 - 30/06/2019
### Update
- [Fix Fix New Structure Zippyshare Download Page, Add mathjs For Evaluate Math String](https://github.com/anasrar/Zippy-DL/commit/2e335136fd4a684389addd57bf99369223c87eee)

## 2.4.0 - 17/06/2019
### Update
- Add Command Downloader
  - GDriveDL https://github.com/ariakm25/GDriveDL
  - ZippyDL https://github.com/anasrar/Zippy-DL
- Forget What Kind Bug That I Have To Fix 😐

## 2.3.5 - 19/05/2019
### Update
- Remove Babel [#2](https://github.com/anasrar/Samehadaku-Lewatin/issues/2)
- Move Changelog In README.md to CHANGELOG.md

## 2.2.5 - 19/05/2019
### Update
- Add New Safelink URL [ njiir.com ]

## 2.2.3 - 28/04/2019
### Update
- Fix Fetch List Anime Use JSON

## 2.1.2 - 15/04/2019
### Fix
- Fix Pagination

## 2.1.1 - 13/04/2019
### Added
- Command From URL Page

## 2.0.1 - 13/04/2019
### Added
- Command Search
- Add Select Next And Prev Page
- Optimizing Code

## 1.2.9 - 23/02/2019
### Added
- Command Save 1 Quality Link As Text File [#1](https://github.com/anasrar/Samehadaku-Lewatin/issues/1)
- Command Page
- Optimizing Code

## 1.2.8 - 22/02/2019
### Added
- Copy to clipboard
- New Safelink URL [ siherp.com ]

[i1]: https://github.com/anasrar/Samehadaku-Lewatin/issues/1

module/dl.js
const _proggers = require('cli-progress'),
    _colors = require('chalk'),
    _fs = require('fs'),
    _$ = require('cheerio'),
    _url = require('url'),
    _https = require('https'),
    _axios = require('axios'),
    _math = require('mathjs')

const querySearch = (q, u) => {
    if (!u) return null;
    q = q.replace(/[\[]/, "\\\[").replace(/[\]]/, "\\\]");
    var regexS = "[\\?&]" + q + "=([^&#]*)";
    var regex = new RegExp(regexS);
    var results = regex.exec(u);
    return results == null ? null : results[1];
}

const clacSize = (a, b) => {
    if (0 == a) return "0 Bytes";
    var c = 1024,
        d = b || 2,
        e = ["Bytes", "KB", "MB", "GB", "TB", "PB", "EB", "ZB", "YB"],
        f = Math.floor(Math.log(a) / Math.log(c));
    return parseFloat((a / Math.pow(c, f)).toFixed(d)) + " " + e[f]
}
const zippyGetLink = async (u) => {
    console.log('⏳  ' + _colors.yellow(`Get Page From : ${u}`))
    const zippy = await _axios({ method: 'GET', url: u }).then(res => res.data).catch(err => false)
    console.log('✅  ' + _colors.green('Done'))
    console.log('⏳  ' + _colors.yellow('Fetch Link Download...'))
    const $ = _$.load(zippy)
    const url = _url.parse($('.flagen').attr('href'), true)
    const urlori = _url.parse(u)
    const key = url.query['key']
    const time = _math.evaluate(/\(([\d\s\+\%]+?)\)/gm.exec($('#dlbutton').next().html())[1])
    const dlurl = urlori.protocol + '//' + urlori.hostname + '/d/' + key + '/' + (time) + '/DOWNLOAD'
    console.log('✅  ' + _colors.green('Done'))
    return dlurl
}
exports.ZippyDL = async (u, cb = () => { }) => {
    const url = await zippyGetLink(u)
    const req = await _https.get(url)
    console.log('🎁  ' + _colors.yellow('Start Download From URL : ' + url))
    console.log('⏳  ' + _colors.yellow('Waiting Server Response...'));
    await req.on('response', res => {
        if (!res.headers['content-disposition']) {
            console.log('🔁  ' + _colors.blue('Server Download Error, Try To Get New Link...'))
            exports.ZippyDL(u, cb)
        } else {
            console.log('✅  ' + _colors.green('Server Response'))
            const size = parseInt(res.headers['content-length'], 10),
                filename = decodeURIComponent(res.headers['content-disposition'].match(/filename\*?=['"]?(?:UTF-\d['"]*)?([^;\r\n"']*)['"]?;?/)[1])
            let currentSize = 0
            console.log('☕  ' + _colors.yellow('Start Downloading File : ' + filename))
            const file = _fs.createWriteStream(filename)
            res.pipe(file)
            const loadbar = new _proggers.Bar({
                format: 'Downloading ' + _colors.green('{bar}') + ' {percentage}% | {current}/{size} | ETA: {eta}s | Speed: {speed}',
                barsize: 25
            }, _proggers.Presets.shades_classic)
            loadbar.start(size, 0, {
                size: clacSize(size, 3),
                current: clacSize(currentSize, 3),
                speed: 0
            })
            res.on('data', c => {
                currentSize += c.length;
                loadbar.increment(c.length, {
                    speed: clacSize(c.length),
                    current: clacSize(currentSize, 3)
                })
            })
            res.on('end', _ => {
                loadbar.stop()
                file.close()
                console.log('✅  ' + _colors.green('Success Download File : ' + filename))
                cb()
            })
            res.on('error', _ => {
                loadbar.stop()
                file.close()
                console.log('❎  ' + _colors.green('Error Download File : ' + filename))
                cb()
            })
        }
    })
}
const gdGetLink = async (url) => {
    let id = querySearch('id', url)
    if (!id) {
        id = /\/file\/d\/(.+)\//gm.exec(url)[1]
    }
    const urlx = `https://drive.google.com/uc?id=${id}&authuser=0&export=download`
    console.log('⏳  ' + _colors.yellow(`Get Page From : ${urlx}`))
    const data = await _axios.post(urlx, {}, {
        headers: {
            'accept-encoding': 'gzip, deflate, br',
            'content-length': 0,
            'Content-Type': 'application/x-www-form-urlencoded;charset=UTF-8',
            'origin': 'https://drive.google.com',
            'user-agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/65.0.3325.181 Safari/537.36',
            'x-client-data': 'CKG1yQEIkbbJAQiitskBCMS2yQEIqZ3KAQioo8oBGLeYygE=',
            'x-drive-first-party': 'DriveWebUi',
            'x-json-requested': 'true'
        }
    }).then(data => { return { success: true, json: JSON.parse(data.data.slice(4)) } }).catch(() => { return { success: false } })
    return data
}
exports.GDriveDL = async (url, cb = () => { }) => {
    const data = await gdGetLink(url)
    console.log('✅  ' + _colors.green('Done'))
    if (!data.success) return console.log('❎  ' + _colors.red('File not Found!'));
    if (data.json.downloadUrl === undefined) return console.log('❎  ' + _colors.red('Link Download LIMIT'));
    console.log('⏳  ' + _colors.yellow('Fetch Link Download...'))
    console.log('✅  ' + _colors.green('Done'))
    console.log('🎁  ' + _colors.yellow('Start Download From URL : ' + data.json.downloadUrl))
    const req = _https.get(data.json.downloadUrl)
    console.log('⏳  ' + _colors.yellow('Waiting Server Response...'))
    req.on('response', (res) => {
        console.log('✅  ' + _colors.green('Server Response'))
        let size = parseInt(data.json.sizeBytes, 10),
            currentSize = 0,
            filename = decodeURIComponent(res.headers['content-disposition'].match(/filename\*?=['"]?(?:UTF-\d['"]*)?([^;\r\n"']*)['"]?;?/)[1])
        console.log('☕  ' + _colors.yellow('Start Downloading File : ' + filename))
        const file = _fs.createWriteStream(filename);
        res.pipe(file)
        const loadbar = new _proggers.Bar({
            format: 'Downloading ' + _colors.green('{bar}') + ' {percentage}% | {current}/{size} | ETA: {eta}s | Speed: {speed}',
            barsize: 25
        }, _proggers.Presets.shades_classic)
        loadbar.start(size, 0, {
            size: clacSize(size, 3),
            current: clacSize(currentSize, 3),
            speed: 0
        })
        res.on('data', c => {
            currentSize += c.length;
            loadbar.increment(c.length, {
                speed: clacSize(c.length),
                current: clacSize(currentSize, 3)
            })
        })
        res.on('end', _ => {
            loadbar.stop()
            file.close()
            console.log('✅  ' + _colors.green('Success Download File : ' + filename))
            cb()
        })
        res.on('error', _ => {
            loadbar.stop()
            file.close()
            console.log('❎  ' + _colors.green('Error Download File : ' + filename))
            cb()
        })
    })
}


module/fetch.js
const axios = require('axios')
const chalk = require('chalk')
const cheerio = require('cheerio')
const clipboard = require('clipboardy')
const fs = require('fs')
const inquirer = require('inquirer')
const urlz = require('url')

let shorturl = ['coeg.in', 'telondasmu.com', 'tetew.info', 'greget.space', 'siherp.com', 'ahexa.com'],
    safeurl = ['njiir.com'],
    safeurl2 = ['eue.siherp.com'],
    safelink2 = ['anjay.info'],
    collect = {};

const querySearch = (q, u) => {
    if (!u) return null;
    q = q.replace(/[\[]/, "\\\[").replace(/[\]]/, "\\\]");
    var regexS = "[\\?&]" + q + "=([^&#]*)";
    var regex = new RegExp(regexS);
    var results = regex.exec(u);
    return results == null ? null : results[1];
}

const getPage = async (url, save = false) => {
    let { data } = await axios.get(url)
    getLinks(data, save)
}

const getList = async (url, save = false) => {
    let arrQuest = []
    let { headers, data } = await axios.get(url)
    let page = parseInt(querySearch('page', url))
    let total = parseInt(headers['x-wp-totalpages'])
    data.forEach(c => {
        arrQuest.push({
            name: c.title.rendered,
            value: c.content.rendered
        },
            new inquirer.Separator(chalk.gray(`└> ${new Date(c.date).toLocaleString('en-us', { month: 'long', year: 'numeric', day: 'numeric' })}`)));
    })
    if (page < total) arrQuest.push({
        name: `Next Page > ${(page + 1)}`,
        value: {
            page: true,
            url: "https://www.samehadaku.tv/wp-json/wp/v2/posts?per_page=14&page=" + (page + 1)
        }
    })
    if (page > 1) arrQuest.push({
        name: `Prev Page < ${(page - 1)}`,
        value: {
            page: true,
            url: "https://www.samehadaku.tv/wp-json/wp/v2/posts?per_page=14&page=" + (page - 1)
        }
    })
    let answer = await inquirer.prompt([{
        type: "list",
        pageSize: (arrQuest.length * 2),
        name: "result",
        message: "Select Please",
        choices: arrQuest
    }]).then(a => a)
    if (answer.result.page === true) {
        getList(answer.result.url, save)
    } else {
        getLinks(answer.result, save)
    }
}

const getLinks = async (html, save = false) => {
    let arrQuest = []
    let stamp = new Date().getTime()
    let $ = cheerio.load(html)

    $('.download-eps').each((i, el) => {
        arrQuest.push({
            name: $(el).prev().text(),
            value: {
                html: $(el).html(),
                type: $(el).prev().text().trim()
            }
        })
    })

    let answer = await inquirer.prompt([{
        type: "list",
        name: "format",
        message: "Select The Format",
        choices: arrQuest
    }]).then(a => a)
    $ = cheerio.load(answer.format.html)

    if (save) {
        $('li').each((i, el) => {
            let reso = $(el).find('strong').text().trim()
            collect[reso] = {};
            $(el).find('a').each((i, el) => {
                let server = $(el).text().trim(),
                    link = $(el).attr('href');
                collect[reso][server] = link
                Pancal(link, save, { reso, server, name: stamp, type: answer.format.type });
            })
        })
    } else {
        arrQuest = []
        $('li').each((i, el) => {
            let reso = $(el).find('strong').text().trim()
            arrQuest.push(new inquirer.Separator(chalk.green(`████ ${reso} ████`)))
            $(el).find('a').each((i, el) => {
                arrQuest.push({
                    name: $(el).text() + ' [' + reso + ']',
                    value: $(el).attr('href')
                })
            })
        })
        let answer2 = await inquirer.prompt([{
            type: "list",
            name: "url",
            message: "Select The Resolution and Server ",
            choices: arrQuest
        }]).then(a => a)
        Pancal(answer2.url)
    }
}

const Pancal = async (url, save = false, config = {}) => {
    let next = !1
    let urlnya = null

    let { data } = await axios.get(url)

    let $ = cheerio.load(data)
    urlnya = $('#splash').find('a[href*="?r=a"]')
    if (urlnya.length) {
        urlnya = $('#splash').find('a[href*="?r=a"]').attr('href');
        urlnya = Buffer.from(querySearch('r', urlnya), 'base64').toString('ascii');
    } else {
        if (safeurl.indexOf(urlz.parse(url).hostname.replace(/^(?:https?:\/\/)?(?:www\.)?/i, "").split('/')[0]) !== -1) {
            urlnya = decodeURIComponent(querySearch('url', url))
            urlnya = decodeURIComponent(Buffer.from(urlnya, 'base64').toString('ascii').substr(2).split('').reverse().join('').substr(2).split('').reverse().join(''))
            let build = '';
            for (let i = 0; i < urlnya.length; i++) {
                build = String.fromCharCode(40 ^ urlnya.charCodeAt(i)) + build;
            }
            urlnya = build
        } else if (safeurl2.indexOf(urlz.parse(url).hostname.replace(/^(?:https?:\/\/)?(?:www\.)?/i, "").split('/')[0]) !== -1) {
            urlnya = decodeURIComponent(querySearch('url', url))
            urlnya = decodeURIComponent(Buffer.from(urlnya, 'base64').toString('ascii'))
        } else if (safelink2.indexOf(urlz.parse(url).hostname.replace(/^(?:https?:\/\/)?(?:www\.)?/i, "").split('/')[0]) !== -1) {
            var params = new URLSearchParams();
            params.append('eastsafelink_id', querySearch('id', url));
            let { data } = await axios.post('https://www.anjay.info/cluster-detection/', params).then(a => a)
            urlnya = data.match(/var a='(https?:\/\/[^ ]*)'/)[1]
            next = false
        } else {
            urlnya = url
        }
    }
    next = [...shorturl, ...safeurl, ...safeurl2, ...safelink2].indexOf(urlz.parse(urlnya).hostname.replace(/^(?:https?:\/\/)?(?:www\.)?/i, "").split('/')[0]) !== -1 ? urlnya : false;
    if (next) {
        Pancal(next, save, config)
    } else {
        if (save) {
            collect[config.reso][config.server] = urlnya
            fs.writeFileSync('./[' + config.type + ']' + config.name + '.txt', JSON.stringify(collect, null, 4))
            console.log(`Proccesing...`)
        } else {
            console.log(`URL : ${chalk.blue(urlnya)}`)
            clipboard.writeSync(urlnya)
            console.log(chalk.green('URL Has Copy On Your Clipboard'))
        }
    }
}

module.exports = {
    getList,
    getPage
}

module/fetchold.js
const request = require('request-promise');
const cheerio = require('cheerio');
const inquirer = require('inquirer');
const chalk = require('chalk');
const urlz = require('url');
const clipboard = require('clipboardy');
const fs = require('fs');
const path = require('path');

let shorturl = ['coeg.in', 'telondasmu.com', 'tetew.info', 'greget.space', 'siherp.com'],
    safeurl = ['njiir.com'],
    safeurl2 = ['eue.siherp.com'],
    collect = {};

const querySearch = (q, u) => {
    if (!u) return null;
    q = q.replace(/[\[]/, "\\\[").replace(/[\]]/, "\\\]");
    var regexS = "[\\?&]" + q + "=([^&#]*)";
    var regex = new RegExp(regexS);
    var results = regex.exec(u);
    return results == null ? null : results[1];
}

const getListSearch = (url, save = false) => {
    let arrQuest = [];
    request(url, (err, res, body) => {
        if (err && res.statusCode !== 200) throw err;

        let $ = cheerio.load(body),
            list = $('body').find('#posts-container>li'),
            prev = $('body').find('li.current').prev(),
            next = $('body').find('li.current').next();

        if (!list.length) {
            console.log(`anime with query ${decodeURIComponent(querySearch('s', url))} has not found`)
        } else {
            list.each((i, el) => {
                arrQuest.push({
                    name: $(el).find('.post-title').text().trim(),
                    value: $(el).find('.post-title>a').attr('href')
                },
                    new inquirer.Separator(chalk.gray(`└> ${$(el).find('.date.meta-item').text().trim()}`)));
                if ((list.length - 1) === i && next.length) arrQuest.push({
                    name: `Next Page > ${next.find('a').text()}`,
                    value: {
                        page: true,
                        url: next.find('a').attr('href')
                    }
                });
                if ((list.length - 1) === i && prev.length) arrQuest.push({
                    name: `Prev Page < ${prev.find('a').text()}`,
                    value: {
                        page: true,
                        url: prev.find('a').attr('href')
                    }
                });
            })
        }

    }).then(() => {
        inquirer.prompt([{
            type: "list",
            pageSize: (arrQuest.length * 2),
            name: "result",
            message: "Select Please",
            choices: arrQuest
        }]).then((a) => {
            if (a.result.page === true) {
                getListSearch(a.result.url, save)
            } else {
                getLinks(a.result, save)
            }
        })
    })
}

const getList = (url, save = false) => {
    let arrQuest = [];
    request(url, (err, res, body) => {
        if (err && res.statusCode !== 200) throw err;

        let $ = cheerio.load(body),
            list = $('#tie-block_1744').find('.posts-items>li'),
            prev = $('#tie-block_1744').find('li.current').prev(),
            next = $('#tie-block_1744').find('li.current').next();

        list.each((i, el) => {
            arrQuest.push({
                name: $(el).find('.post-title').text().trim(),
                value: $(el).find('.post-title>a').attr('href')
            },
                new inquirer.Separator(chalk.gray(`└> ${$(el).find('.date.meta-item').text().trim()}`)));
            if ((list.length - 1) === i && next.length) arrQuest.push({
                name: `Next Page > ${next.find('a').text()}`,
                value: {
                    page: true,
                    url: next.find('a').attr('href')
                }
            });
            if ((list.length - 1) === i && prev.length) arrQuest.push({
                name: `Prev Page < ${prev.find('a').text()}`,
                value: {
                    page: true,
                    url: prev.find('a').attr('href')
                }
            });
        });
    }).then(() => {
        inquirer.prompt([{
            type: "list",
            pageSize: (arrQuest.length * 2),
            name: "result",
            message: "Select Please",
            choices: arrQuest
        }]).then((a) => {
            if (a.result.page === true) {
                getList(a.result.url, save)
            } else {
                getLinks(a.result, save)
            }
        })
    })
}

const getLinks = (url, save = false) => {
    let arrQuest = [];

    request(url, (err, res, body) => {
        if (err && res.statusCode !== 200) throw err;

        let $ = cheerio.load(body);

        $('.download-eps').each((i, el) => {
            arrQuest.push({
                name: $(el).prev().text(),
                value: {
                    html: $(el).html(),
                    type: $(el).prev().text().trim()
                }
            });
        });

    }).then(() => {

        inquirer.prompt([{
            type: "list",
            name: "format",
            message: "Select The Format",
            choices: arrQuest
        }]).then(a => {
            let $ = cheerio.load(a.format.html);

            if (save) {
                $('li').each((i, el) => {
                    let reso = $(el).find('strong').text().trim();
                    collect[reso] = {};
                    $(el).find('a').each((i, el) => {
                        let server = $(el).text().trim(),
                            link = $(el).attr('href');
                        collect[reso][server] = link;
                        Pancal(link, save, { reso, server, name: path.basename(url, '.html'), type: a.format.type })
                    })
                })
            } else {
                arrQuest = [];
                $('li').each((i, el) => {
                    let reso = $(el).find('strong').text().trim();
                    arrQuest.push(new inquirer.Separator(chalk.green(`████ ${reso} ████`)));
                    $(el).find('a').each((i, el) => {
                        arrQuest.push({
                            name: $(el).text() + ' [' + reso + ']',
                            value: $(el).attr('href')
                        });
                    });
                });

                inquirer.prompt([{
                    type: "list",
                    name: "url",
                    message: "Select The Resolution and Server ",
                    choices: arrQuest
                }]).then(a => {
                    Pancal(a.url)
                })
            }


        })
    })
}

const Pancal = (url, save = false, config = {}) => {
    let next = !1,
        urlnya;

    request(url, (err, res, body) => {
        if (err && res.statusCode !== 200) throw err;

        let $ = cheerio.load(body);
        urlnya = $('#splash').find('a[href*="?r=a"]');
        if (urlnya.length) {
            urlnya = $('#splash').find('a[href*="?r=a"]').attr('href');
            urlnya = Buffer.from(querySearch('r', urlnya), 'base64').toString('ascii');
            next = [...shorturl, ...safeurl, ...safeurl2].indexOf(urlz.parse(urlnya).hostname.replace(/^(?:https?:\/\/)?(?:www\.)?/i, "").split('/')[0]) !== -1 ? urlnya : false;
        } else {
            if (safeurl.indexOf(urlz.parse(url).hostname.replace(/^(?:https?:\/\/)?(?:www\.)?/i, "").split('/')[0]) !== -1) {
                urlnya = decodeURIComponent(querySearch('url', url))
                urlnya = decodeURIComponent(Buffer.from(urlnya, 'base64').toString('ascii').substr(2).split('').reverse().join('').substr(2).split('').reverse().join(''))
                let build = '';
                for (let i = 0; i < urlnya.length; i++) {
                    build = String.fromCharCode(40 ^ urlnya.charCodeAt(i)) + build;
                }
                urlnya = build
                next = [...shorturl, ...safeurl, ...safeurl2].indexOf(urlz.parse(urlnya).hostname.replace(/^(?:https?:\/\/)?(?:www\.)?/i, "").split('/')[0]) !== -1 ? urlnya : false;
            } else if (safeurl2.indexOf(urlz.parse(url).hostname.replace(/^(?:https?:\/\/)?(?:www\.)?/i, "").split('/')[0]) !== -1) {
                urlnya = decodeURIComponent(querySearch('url', url))
                urlnya = decodeURIComponent(Buffer.from(urlnya, 'base64').toString('ascii'))
                next = [...shorturl, ...safeurl, ...safeurl2].indexOf(urlz.parse(urlnya).hostname.replace(/^(?:https?:\/\/)?(?:www\.)?/i, "").split('/')[0]) !== -1 ? urlnya : false;
            } else {
                urlnya = url
            }
        }

    }).then(() => {
        if (next) {
            Pancal(next, save, config)
        } else {
            if (save) {
                collect[config.reso][config.server] = urlnya;
                fs.writeFileSync('./[' + config.type + ']' + config.name + '.txt', JSON.stringify(collect, null, 4));
            } else {
                console.log(`URL : ${chalk.blue(urlnya)}`);
                clipboard.writeSync(urlnya);
                console.log(chalk.green('URL Has Copy On Your Clipboard'))
            }
        }
    })
}

const getListJSON = (url, save = false) => {
    let arrQuest = []
    request(url, (err, res, body) => {
        if (err && res.statusCode !== 200) throw err;

        let page = parseInt(querySearch('page', url)),
            total = parseInt(res.headers['x-wp-totalpages'])

        let list = JSON.parse(body)
        list.forEach(c => {
            arrQuest.push({
                name: c.title.rendered,
                value: c.content.rendered
            },
                new inquirer.Separator(chalk.gray(`└> ${new Date(c.date).toLocaleString('en-us', { month: 'long', year: 'numeric', day: 'numeric' })}`)));
        })
        if (page < total) arrQuest.push({
            name: `Next Page > ${(page + 1)}`,
            value: {
                page: true,
                url: "https://www.samehadaku.tv/wp-json/wp/v2/posts?per_page=14&page=" + (page + 1)
            }
        })
        if (page > 1) arrQuest.push({
            name: `Prev Page < ${(page - 1)}`,
            value: {
                page: true,
                url: "https://www.samehadaku.tv/wp-json/wp/v2/posts?per_page=14&page=" + (page - 1)
            }
        });
    }).then(() => {
        inquirer.prompt([{
            type: "list",
            pageSize: (arrQuest.length * 2),
            name: "result",
            message: "Select Please",
            choices: arrQuest
        }]).then((a) => {
            if (a.result.page === true) {
                getListJSON(a.result.url, save)
            } else {
                getLinksHTML(a.result, save)
            }
        })
    })
}

const getLinksHTML = (html, save = false) => {
    let arrQuest = [],
        stamp = new Date().getTime(),
        $ = cheerio.load(html)

    $('.download-eps').each((i, el) => {
        arrQuest.push({
            name: $(el).prev().text(),
            value: {
                html: $(el).html(),
                type: $(el).prev().text().trim()
            }
        })
    })

    inquirer.prompt([{
        type: "list",
        name: "format",
        message: "Select The Format",
        choices: arrQuest
    }]).then(a => {
        let $ = cheerio.load(a.format.html);

        if (save) {
            $('li').each((i, el) => {
                let reso = $(el).find('strong').text().trim();
                collect[reso] = {};
                $(el).find('a').each((i, el) => {
                    let server = $(el).text().trim(),
                        link = $(el).attr('href');
                    collect[reso][server] = link;
                    Pancal(link, save, { reso, server, name: stamp, type: a.format.type })
                })
            })
        } else {
            arrQuest = [];
            $('li').each((i, el) => {
                let reso = $(el).find('strong').text().trim();
                arrQuest.push(new inquirer.Separator(chalk.green(`████ ${reso} ████`)));
                $(el).find('a').each((i, el) => {
                    arrQuest.push({
                        name: $(el).text() + ' [' + reso + ']',
                        value: $(el).attr('href')
                    });
                });
            });

            inquirer.prompt([{
                type: "list",
                name: "url",
                message: "Select The Resolution and Server ",
                choices: arrQuest
            }]).then(a => {
                Pancal(a.url)
            })
        }


    })
}

module.exports = {
    querySearch,
    getListSearch,
    getList,
    getLinks,
    Pancal,
    getListJSON,
    getLinksHTML
}

module/app.js
const { getPage, getList } = require('./fetch');
const { ZippyDL, GDriveDL } = require('./dl');
const chalk = require('chalk');
const minimist = require('minimist');
const request = require('request-promise');
const semver = require('semver')

let url = "https://www.samehadaku.tv/",
    version = require('../package.json').version;

console.log(`${chalk.yellow(`╔═╗╔═╗╔╦╗╔═╗╦ ║╔═╗╔╦╗╔═╗╦╔═║ ╦  ╦  ╔═╗╦ ╦╔═╗╔╦╗╦╔╗║
╚═╗╠═╣║║║║╣ ╠═╣╠═╣ ║║╠═╣╠╩╗║ ║  ║  ║╣ ║║║╠═╣ ║ ║║║║
╚═╝║ ║╩ ╩╚═╝║ ╩║ ║═╩╝║ ║╩ ╩╚═╝  ╚═╝╚═╝╚╩╝╩ ╩ ╩ ╩║╚╝`)}
===============================================================
Version    : ${version}
Repository : https://github.com/anasrar/Samehadaku-Lewatin
===============================================================`);

(async () => {
    await request("https://raw.githubusercontent.com/anasrar/Samehadaku-Lewatin/master/package.json", (err, res, body) => {
        if (err && res.statusCode !== 200) throw err;
        const newVersion = JSON.parse(body).version
        if (semver.lt(version, newVersion)) {
            console.log(`${chalk.yellow(`We Have New Version : ${newVersion}`)}`)
        }
    })

    const args = minimist(process.argv.slice(2));
    let cmd = args._[0] || 'help';

    if (args.help || args.h) {
        cmd = 'help'
    }
    switch (cmd) {
        case 'list':
            getList(url + 'wp-json/wp/v2/posts?per_page=14&page=1', (args.s === true || args.save === true));
            break;
        case 'page':
            let page = args._[1] || 1;
            getList(url + 'wp-json/wp/v2/posts?per_page=14&page=' + page, (args.s === true || args.save === true));
            break;
        case 'from':
            let laman = args._[1] || url;
            if (laman == url) {
                getListJSON(url + 'wp-json/wp/v2/posts?per_page=14&page=1', (args.s === true || args.save === true));
            } else {
                getPage(laman, (args.s === true || args.save === true));
            }
            break;
        case 'help':
            console.log(
                `Usage: samehadaku <command> [options]

where <command> is one of:
    list, page, help, from, zippydl, gdrivedl

options: [-s|--save] for save link to text file with json format`
            );
            break;
        case 'zippydl':
            url = args._[1]
            if (!url) {
                console.log('Please insert url, e.g : samehadaku zippydl "https://xxx.zippyshare.com/v/xxx/file.html"')
                break;
            }
            ZippyDL(url)
            break;
        case 'gdrivedl':
            url = args._[1]
            if (!url) {
                console.log('Please insert url, e.g : samehadaku gdrive "https://drive.google.com/open?id=xxx" or "https://drive.google.com/file/d/xxx/view"')
                break;
            }
            GDriveDL(url)
            break;
        default:
            console.error(`"${cmd}" is not a valid command!`)
            break;
    }

})();

// console.log(args)